### PR TITLE
[move-only] Emit an error if we /ever/ partially consume a noncopyable type.

### PR DESCRIPTION
**Pull request by gottesmm**
The reason why I am doing this is that this was not part of the original
evolution proposal (it was called an extension) and after some discussion it was
realized that partial consumption would benefit from discussion on the forums.

rdar://111353459

----

I also needed to add a small optimization to eliminate some iterated copies that are inserted by SILGen since otherwise the diagnostic gets borked. It works by just walking def->uses and is very conservative by: 1. Only pattern matching a very exact sequence. 2. Making sure that all visited iterative copies are in the same block and do not have any side effects in the sequence of instructions.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

File: lib/SILOptimizer/Mandatory/MoveOnlyUtils.cpp [hunk 1]
```diff
@@ -63,6 +63,8 @@
 using namespace swift;
 using namespace swift::siloptimizer;
 
+#pragma clang optimize off
```
Comment (jckarter) [inline lib/SILOptimizer/Mandatory/MoveOnlyUtils.cpp:66]: Did you mean to leave this in?

**Review by jckarter — APPROVED** [review]

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

**Pull request merged**